UpdateUserGroups/handler.ts

```diff
@@ -0,0 +1,381 @@
+import { RequiredBodyPayloadMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_body_payload";
+import { RequiredParamMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_param";
+import {
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "io-functions-commons/dist/src/utils/request_middleware";
+import {
+  checkSourceIpForHandler,
+  clientIPAndCidrTuple as ipTuple
+} from "io-functions-commons/dist/src/utils/source_ip_check";
+import {
+  IResponseErrorInternal,
+  IResponseErrorNotFound,
+  IResponseErrorValidation,
+  IResponseSuccessJson,
+  ResponseErrorNotFound,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { EmailAddress } from "../generated/definitions/EmailAddress";
+import { GroupCollection } from "../generated/definitions/GroupCollection";
+import { UserGroupsPayload } from "../generated/definitions/UserGroupsPayload";
+import {
+  getApiClient,
+  IAzureApimConfig,
+  IServicePrincipalCreds
+} from "../utils/apim";
+import { groupContractToApiGroup } from "../utils/conversions";
+import { genericInternalErrorHandler } from "../utils/errorHandler";
+
+type IGetSubscriptionKeysHandlerResponseError =
+  | IResponseErrorNotFound
+  | IResponseErrorInternal
+  | IResponseErrorValidation;
+type IGetSubscriptionKeysHandler = (
+  context: Context,
+  auth: IAzureApiAuthorization,
+  clientIp: ClientIp,
+  userAttributes: IAzureUserAttributes,
+  email: EmailAddress,
+  userGroupsPayload: UserGroupsPayload
+) => Promise<
+  | IResponseSuccessJson<GroupCollection>
+  | IGetSubscriptionKeysHandlerResponseError
+>;
+
+function getGroups(
+  apimClient: ApiManagementClient,
+  apimResourceGroup: string,
+  apim: string
+): TaskEither<Error, ReadonlyArray<GroupContract>> {
+  return tryCatch(async () => {
+    // tslint:disable-next-line:readonly-array no-let
+    const groupList: GroupContract[] = [];
+    const groupListResponse = await apimClient.group.listByService(
+      apimResourceGroup,
+      apim
+    );
+    groupList.push(...groupListResponse);
+    // tslint:disable-next-line:no-let
+    let nextLink = groupListResponse.nextLink;
+    while (nextLink) {
+      const nextGroupList = await apimClient.group.listByServiceNext(nextLink);
+      groupList.push(...nextGroupList);
+      nextLink = nextGroupList.nextLink;
+    }
+    return groupList;
+  }, toError);
+}
+
+function getUserGroups(
+  apimClient: ApiManagementClient,
+  apimResourceGroup: string,
+  apim: string,
+  userName: string
+): TaskEither<Error, ReadonlyArray<GroupContract>> {
+  return tryCatch(async () => {
+    // tslint:disable-next-line:readonly-array no-let
+    const groupList: GroupContract[] = [];
+    const groupListResponse = await apimClient.userGroup.list(
+      apimResourceGroup,
+      apim,
+      userName
+    );
+    groupList.push(...groupListResponse);
+    // tslint:disable-next-line:no-let
+    let nextLink = groupListResponse.nextLink;
+    while (nextLink) {
```

Comment:
not strictly needed for this pr but when you want to accumulate vaules another option is to use [unfoldr](https://github.com/gcanti/fp-ts/blob/1.12.0/src/Array.ts#L227) which is a do ... while that takes a function as an input that returns an Option (some -> put the value into the array and continue, none -> stop the loop)

Reply:
Thanks for the information! By the way, here we are accumulating values that come from a remote call using `await`; is there an equivalent method that allows using an asynchronous function? It seems I didn't see any in the docs...